sql/core/src/test/scala/org/apache/spark/sql/execution/SelectedFieldSuite.scala

```diff
@@ -0,0 +1,415 @@
+  //  |    |-- value: array (valueContainsNull = true)
+  //  |    |    |-- element: struct (containsNull = false)
+  //  |    |    |    |-- field1: struct (nullable = true)
+  //  |    |    |    |    |-- subfield1: integer (nullable = true)
+  //  |    |    |    |    |-- subfield2: integer (nullable = true)
+  //  |-- col7: array (nullable = true)
+  //  |    |-- element: struct (containsNull = true)
+  //  |    |    |-- field1: integer (nullable = false)
+  //  |    |    |-- field2: struct (nullable = true)
+  //  |    |    |    |-- subfield1: integer (nullable = false)
+  //  |    |    |-- field3: array (nullable = true)
+  //  |    |    |    |-- element: struct (containsNull = true)
+  //  |    |    |    |    |-- subfield1: integer (nullable = false)
+  //  |-- col8: array (nullable = true)
+  //  |    |-- element: struct (containsNull = true)
+  //  |    |    |-- field1: array (nullable = false)
+  //  |    |    |    |-- element: integer (containsNull = false)
+  private val schema =
+    StructType(
+      StructField("col1", StringType, nullable = false) ::
+      innerSchemaCol1 ::
+      StructField("col3", ArrayType(StructType(
+        StructField("field1", StructType(
+         StructField("subfield1", IntegerType, nullable = false) ::
+         StructField("subfield2", IntegerType) :: Nil)) ::
+       StructField("field2", MapType(StringType, IntegerType, valueContainsNull = false))
+          :: Nil), containsNull = false), nullable = false) ::
+     StructField("col4", MapType(StringType, StructType(
+       StructField("field1", StructType(
+         StructField("subfield1", IntegerType, nullable = false) ::
+         StructField("subfield2", IntegerType) :: Nil)) ::
+       StructField("field2", MapType(StringType, IntegerType, valueContainsNull = false))
+         :: Nil), valueContainsNull = false), nullable = false) ::
+     StructField("col5", ArrayType(MapType(StringType, StructType(
+       StructField("field1", StructType(
+         StructField("subfield1", IntegerType) ::
+         StructField("subfield2", IntegerType) :: Nil)) :: Nil), valueContainsNull = false))) ::
+     StructField("col6", MapType(StringType, ArrayType(StructType(
+       StructField("field1", StructType(
+         StructField("subfield1", IntegerType) ::
+         StructField("subfield2", IntegerType) :: Nil)) :: Nil), containsNull = false))) ::
+     StructField("col7", ArrayType(StructType(
+       StructField("field1", IntegerType, nullable = false) ::
+       StructField("field2", StructType(
+         StructField("subfield1", IntegerType, nullable = false) :: Nil)) ::
+       StructField("field3", ArrayType(StructType(
+         StructField("subfield1", IntegerType, nullable = false) :: Nil))) :: Nil))) ::
+     StructField("col8", ArrayType(StructType(
+       StructField("field1", ArrayType(IntegerType, containsNull = false), nullable = false)
+         :: Nil))) :: Nil)
+
+  private val testRelation = LocalRelation(schema.toAttributes)
```

Comment:
Shall we have multiple relations with split schema for each? It wouldn't necessarily have one big deep nested schema which makes hard to read.